docs/bounded_dynamism.md

```diff
@@ -0,0 +1,85 @@
+
+Currently, StableHLO dialect is using the MLIR core ranked tensor type to encode bounds. It should be noted that there is a plan to introduce a custom StableHLO type in the future that could natively support bounds along with custom pretty printing format. There will be a separate RFC describing all the details. Also, the proposal will follow StableHLO backward compatibility policies so it is safe to use this type currently.
+
+
+## (P2) StableHLO op semantics with bounded operands or results
+
+All ops that support dynamic operands or results can have bounds specified for them. However, the result types needs to be compatible with the inferrable result types using the operands and attributes. This allows result types to be more generic or specific as long as it is compatible with the inferred type.
+
+
+## (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and dimension index as operands and returns the actual size of the dimension at runtime as `i32` type scalar.
+
+The following example returns the size of the result after concatenating input that has up to `16` elements with self and returns the actual runtime size of the concatenation result.
+
+
+```
+func.func @self_concat_size(%data: tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>) -> tensor<i32> {
+  %concat = "stablehlo.concatenate"(%data, %data) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>, 
+       tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>)
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>
+
+  %result = "stablehlo.get_dimension_size"(%concat) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>) -> tensor<i32>
+  
+  func.return %result : tensor<i32>
+}
+```
+
+
+The `set_dimension_size` op takes a tensor, runtime size and dimension index as operands and returns a tensor whose logical size of the particular dimension is set to the specified value. This size needs to be less than or equal to the static size, if available. 
+
+In the following example, set dimension size is used to set the logical size of the first dimension. With input `[1, 2, 3, 4]`, it returns `3` with batch\_size equal to `2` but returns `6` with batch\_size `3`.
+
+
+```
+func.func @dynamic_sum(%data: tensor<4xi32>, %batch_size: tensor<i32>) -> tensor<i32> {
+  %dim = stablehlo.constant dense<0> : tensor<i32>
+  %dynamic_data =  "stablehlo.set_dimension_size"(%data, %batch_size, %dim)
+    : (tensor<4xi32>, tensor<i32>, tensor<i32>) 
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>
```

Comment:
Clarify where the bound `4` is coming from. Does the operand to `set_dimension_size` need to be a statically sized tensor (I assume that is the case).